### PR TITLE
Do not repartition in local exchange for partial row number node

### PR DESCRIPTION
**Pull request by feilong-liu**
For partial row number node, we do not need all data to be partitioned, hence follow the partial topn row number node for local exchange.

### Test plan - (Please fill in how you tested your changes)

Test end to end for query `explain (type distributed) select * from (select *, row_number() over (partition by orderkey) <= 1 as keep from lineitem) where keep;`
Without the change here, a hash local exchange is added.
```
Fragment 2 [SOURCE]                                                                                                                           >
     Output layout: [orderkey, partkey, suppkey, linenumber, quantity, extendedprice, discount, tax, returnflag, linestatus, shipdate, commitda>
     Output partitioning: HASH [orderkey][$hashvalue_138]                                                                                      >
     Stage Execution Strategy: UNGROUPED_EXECUTION                                                                                             >
     - RowNumberPartial[partition by (orderkey), limit = 1][$hashvalue_138] => [orderkey:bigint, partkey:bigint, suppkey:bigint, linenumber:int>
             Estimates: {rows: 15432 (2.38MB), cpu: 31174590.55, memory: 2497710.55, network: 0.00}                                            >
             row_number_136 := row_number()                                                                                                    >
         - LocalExchange[HASH][$hashvalue_138] (orderkey) => [orderkey:bigint, partkey:bigint, suppkey:bigint, linenumber:integer, quantity:dou>
                 Estimates: {rows: 60175 (9.29MB), cpu: 28676880.00, memory: 0.00, network: 0.00}                                              >
             - ScanProject[table = TableHandle {connectorId='hive', connectorHandle='HiveTableHandle{schemaName=tpch, tableName=lineitem, analy>
                     Estimates: {rows: 60175 (9.29MB), cpu: 9197910.00, memory: 0.00, network: 0.00}/{rows: 60175 (9.29MB), cpu: 18937395.00, m>
                     $hashvalue_139 := combine_hash(BIGINT'0', COALESCE($operator$hash_code(orderkey), BIGINT'0')) (1:113)                     >
                     LAYOUT: tpch.lineitem{}                                                                                                   >
                     quantity := quantity:double:4:REGULAR (1:113)                                                                             >
                     receiptdate := receiptdate:date:12:REGULAR (1:113)                                                                        >
                     shipinstruct := shipinstruct:varchar(25):13:REGULAR (1:113)                                                               >
                     discount := discount:double:6:REGULAR (1:113)                                                                             >
                     orderkey := orderkey:bigint:0:REGULAR (1:113)                                                                             >
                     shipdate := shipdate:date:10:REGULAR (1:113)                                                                              >
                     returnflag := returnflag:varchar(1):8:REGULAR (1:113)                                                                     >
                     suppkey := suppkey:bigint:2:REGULAR (1:113)                                                                               >
                     shipmode := shipmode:varchar(10):14:REGULAR (1:113)                                                                       >
                     partkey := partkey:bigint:1:REGULAR (1:113)                                                                               >
                     linenumber := linenumber:int:3:REGULAR (1:113)                                                                            >
                     extendedprice := extendedprice:double:5:REGULAR (1:113)                                                                   >
                     tax := tax:double:7:REGULAR (1:113)                                                                                       >
                     comment := comment:varchar(44):15:REGULAR (1:113)                                                                         >
                     commitdate := commitdate:date:11:REGULAR (1:113)                                                                          >
                     linestatus := linestatus:varchar(1):9:REGULAR (1:113)  
```
With the change here, no local exchange added
```
Fragment 2 [SOURCE]                                                                                                                           >
     Output layout: [orderkey, partkey, suppkey, linenumber, quantity, extendedprice, discount, tax, returnflag, linestatus, shipdate, commitda>
     Output partitioning: HASH [orderkey][$hashvalue_138]                                                                                      >
     Stage Execution Strategy: UNGROUPED_EXECUTION                                                                                             >
     - RowNumberPartial[partition by (orderkey), limit = 1][$hashvalue_138] => [orderkey:bigint, partkey:bigint, suppkey:bigint, linenumber:int>
             Estimates: {rows: 15432 (2.38MB), cpu: 21435105.55, memory: 2497710.55, network: 0.00}                                            >
             row_number_136 := row_number()                                                                                                    >
         - ScanProject[table = TableHandle {connectorId='hive', connectorHandle='HiveTableHandle{schemaName=tpch, tableName=lineitem, analyzePa>
                 Estimates: {rows: 60175 (9.29MB), cpu: 9197910.00, memory: 0.00, network: 0.00}/{rows: 60175 (9.29MB), cpu: 18937395.00, memor>
                 $hashvalue_138 := combine_hash(BIGINT'0', COALESCE($operator$hash_code(orderkey), BIGINT'0')) (1:85)                          >
                 LAYOUT: tpch.lineitem{}                                                                                                       >
                 suppkey := suppkey:bigint:2:REGULAR (1:113)                                                                                   >
                 quantity := quantity:double:4:REGULAR (1:113)                                                                                 >
                 shipinstruct := shipinstruct:varchar(25):13:REGULAR (1:113)                                                                   >
                 commitdate := commitdate:date:11:REGULAR (1:113)                                                                              >
                 linenumber := linenumber:int:3:REGULAR (1:113)                                                                                >
                 returnflag := returnflag:varchar(1):8:REGULAR (1:113)                                                                         >
                 discount := discount:double:6:REGULAR (1:113)                                                                                 >
                 partkey := partkey:bigint:1:REGULAR (1:113)                                                                                   >
                 tax := tax:double:7:REGULAR (1:113)                                                                                           >
                 shipmode := shipmode:varchar(10):14:REGULAR (1:113)                                                                           >
                 extendedprice := extendedprice:double:5:REGULAR (1:113)                                                                       >
                 comment := comment:varchar(44):15:REGULAR (1:113)                                                                             >
                 shipdate := shipdate:date:10:REGULAR (1:113)                                                                                  >
                 linestatus := linestatus:varchar(1):9:REGULAR (1:113)                                                                         >
                 orderkey := orderkey:bigint:0:REGULAR (1:113)                                                                                 >
                 receiptdate := receiptdate:date:12:REGULAR (1:113) 
```

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/RowNumberNode.java [hunk 1]
```diff
@@ -126,7 +126,7 @@ public Optional<Integer> getMaxRowCountPerPartition()
     }
 
     @JsonProperty
-    public boolean getPartial()
+    public boolean isPartial()
```
Comment (feilong-liu) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/RowNumberNode.java:129]: Follow the TopNRowNumber node, rename to isPartial

File: presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java [hunk 2]
```diff
@@ -793,7 +793,7 @@ public Void visitRowNumber(RowNumberNode node, Void context)
             }
 
             NodeRepresentation nodeOutput = addNode(node,
-                    "RowNumber",
+                    format("RowNumber%s", node.isPartial() ? "Partial" : ""),
```
Comment (feilong-liu) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java:796]: Mark the node as RowNumberPartial for partial row number node in plan printer

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddLocalExchanges.java [hunk 3]
```diff
@@ -509,7 +509,11 @@ else if (!node.getDistinctVariables().containsAll(property.getColumns())) {
         public PlanWithProperties visitRowNumber(RowNumberNode node, StreamPreferredProperties parentPreferences)
         {
             // row number requires that all data be partitioned
-            StreamPreferredProperties requiredProperties = parentPreferences.withDefaultParallelism(session).withPartitioning(node.getPartitionBy());
+            StreamPreferredProperties requiredProperties = parentPreferences.withDefaultParallelism(session);
+            // final row number requires that all data be partitioned
+            if (!node.isPartial()) {
+                requiredProperties = requiredProperties.withPartitioning(node.getPartitionBy());
+            }
```
Comment (feilong-liu) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddLocalExchanges.java:516]: Follow what we did for TopNRowNumber node.

**Comment by kaikalur** [comment]
I wonder if adding local RR shuffle makes it even better. It's relatively cheap. Good to run some tests and if looks good, add it.

**Review by pranjalssh — APPROVED** [review]

**Pull request merged**